### PR TITLE
feat(x/data): attest to multiple pieces of data

### PR DESCRIPTION
**Pull request by ryanchristo**
## Description

Closes: #961

This pull request updates `Attest` so that one attestor can attest to multiple pieces of data. This pull request also updates the response to include new attestation entries (attesting to the same piece of data is a no-op).

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/data/msgs.go [hunk 1]
```diff
@@ -39,26 +39,25 @@ func (m MsgAnchor) GetSignBytes() []byte {
...
-		if _, err := sdk.AccAddressFromBech32(addr); err != nil {
-			return sdkerrors.ErrInvalidAddress.Wrap(err.Error())
-		}
+	if _, err := sdk.AccAddressFromBech32(m.Attestor); err != nil {
+		return sdkerrors.ErrInvalidAddress.Wrap(err.Error())
 	}
-	if m.Hash == nil {
-		return sdkerrors.ErrInvalidRequest.Wrap("hash cannot be empty")
+
+	for _, hash := range m.Hashes {
```
Comment (technicallyty) [inline x/data/msgs.go:46]: we should also check that `len(m.Hashes) != 0`

**Review by technicallyty — APPROVED** [review]
pre-approving, need 1 more check in validate basic

File: x/data/msgs.go (hunk 1 above)
Comment (ryanchristo) [reply x/data/msgs.go:50]: Added in https://github.com/regen-network/regen-ledger/pull/971/commits/0681137f97065067f952e43d4c70f9295c05d2ad

File: x/data/server/testsuite/suite.go [hunk 2]
```diff
@@ -35,6 +38,7 @@ func (s *IntegrationTestSuite) SetupSuite() {
 
 	s.fixture = s.fixtureFactory.Setup()
 	s.ctx = s.fixture.Context()
+	s.sdkCtx = s.ctx.(types.Context).WithBlockTime(time.Now())
```
Comment (aaronc) [inline x/data/server/testsuite/suite.go:41]: we need to update `s.ctx` too as in line 93, not sure `time.Now` is smart for tests

**Review by aaronc — APPROVED** [review]
Minor issue to address in tests, pre-approving

File: proto/regen/data/v1/events.proto [hunk 3]
```diff
@@ -6,16 +6,16 @@ option go_package = "github.com/regen-network/regen-ledger/x/data";
...
 message EventAnchor {
-  // iri is the data IRI
+  // iri is the IRI of the data anchored on chain.
   string iri = 1;
 }
 
 // EventAttest is an event emitted when data is attested to on chain.
 message EventAttest {
-  // iri is the data IRI
+  // iri is the IRI of the data attested to.
```
Comment (robert-zaremba) [inline proto/regen/data/v1/events.proto:15]:
```suggestion
  // IRI of the data attested to.
```

File: proto/regen/data/v1/tx.proto [hunk 4]
```diff
@@ -54,42 +53,45 @@ service Msg {
...
+  // veracity of the data referenced by the IRI. The precise meaning of this may
   // vary depending on the underlying data.
-  repeated string attestors = 1;
+  string attestor = 1;
 
-  // hash is the hash-based identifier for the anchored content. Only RDF graph
-  // data can be signed as its data model is intended to specifically convey
-  // semantic meaning.
-  ContentHash.Graph hash = 2;
+  // hashes are the content hashes for anchored data. Only RDF graph data can be
```
Comment (robert-zaremba) [inline proto/regen/data/v1/tx.proto:82]:
```suggestion
  // hashes of the anchored data content. Only RDF graph data can be
```

File: proto/regen/data/v1/types.proto [hunk 5]
```diff
@@ -151,20 +151,25 @@ message ContentEntry {
   // hash is the content hash of the anchored data.
   ContentHash hash = 1;
 
-  // iri is the content IRI of the anchored data.
+  // iri is the IRI of the anchored data.
   string iri = 2;
 
-  // timestamp is the Timestamp of the anchored data.
+  // timestamp is the timestamp of the anchored data.
```
Comment (robert-zaremba) [inline proto/regen/data/v1/types.proto:157]:
```suggestion
  // timestamp of the anchored data.
```

File: proto/regen/data/v1/types.proto [hunk 6]
```diff
@@ -151,20 +151,25 @@ message ContentEntry {
...
   google.protobuf.Timestamp timestamp = 3;
 }
 
-// AttestorEntry is a attestor entry wrapping a attestor address and timestamp
+// AttestorEntry is an attestor entry including the attestor, the IRI, and the
+// timestamp.
 message AttestorEntry {
-  // attestor is the address of the attestor.
-  string attestor = 1;
+  // iri is the IRI of the data attested to.
```
Comment (robert-zaremba) [inline proto/regen/data/v1/types.proto:164]:
```suggestion
  // IRI of the data attested to.
```

File: proto/regen/data/v1/types.proto [hunk 7]
```diff
@@ -151,20 +151,25 @@ message ContentEntry {
...
+  string iri = 1;
 
-  // timestamp is the Timestamp of the attestation.
-  google.protobuf.Timestamp timestamp = 2;
+  // attestor is the address of the account that has attested to the veracity of
+  // the data.
+  string attestor = 2;
+
+  // timestamp is the timestamp at which the data was attested to.
```
Comment (robert-zaremba) [inline proto/regen/data/v1/types.proto:171]:
```suggestion
  // timestamp at which the data was attested to.
```

File: proto/regen/data/v1/events.proto (hunk 3 above)
Comment (ryanchristo) [reply proto/regen/data/v1/events.proto:15]: This is not the convention for writing comments. We always start with the name of the field.

File: x/data/client/tx.go [hunk 8]
```diff
@@ -88,24 +90,31 @@ func MsgAttestCmd() *cobra.Command {
...
-			if len(iri) == 0 {
-				return sdkerrors.ErrInvalidRequest.Wrap("iri is required")
-			}
-
 			attestor := clientCtx.GetFromAddress()
-			content, err := data.ParseIRI(iri)
-			if err != nil {
-				return sdkerrors.ErrInvalidRequest.Wrapf("invalid iri: %s", err.Error())
+
+			if len(args[0]) == 0 {
```
Comment (robert-zaremba) [inline x/data/client/tx.go:95]: I think this check is not needed - Corba is already doing it through line 86

File: proto/regen/data/v1/tx.proto (hunk 4 above)
Comment (ryanchristo) [reply proto/regen/data/v1/tx.proto:82]: This is not the convention for writing comments. We always start with the name of the field.

File: x/data/client/tx.go [hunk 9]
```diff
@@ -88,24 +90,31 @@ func MsgAttestCmd() *cobra.Command {
...
-				return sdkerrors.ErrInvalidRequest.Wrap("can only attest to graph data types")
+
+			var hashes []*data.ContentHash_Graph
+
+			iris := strings.Split(args[0], ",")
+
+			for i := range iris {
+				content, err := data.ParseIRI(iris[i])
+				if err != nil {
+					return sdkerrors.ErrInvalidRequest.Wrapf("invalid iri: %s", err.Error())
```
Comment (robert-zaremba) [inline x/data/client/tx.go:106]: how about adding index to the error message?

File: proto/regen/data/v1/types.proto (hunk 5 above)
Comment (ryanchristo) [reply proto/regen/data/v1/types.proto:157]: This is not the convention for writing comments. We always start with the name of the field.

File: proto/regen/data/v1/types.proto (hunk 6 above)
Comment (ryanchristo) [reply proto/regen/data/v1/types.proto:164]: This is not the convention for writing comments. We always start with the name of the field.

File: proto/regen/data/v1/types.proto (hunk 7 above)
Comment (ryanchristo) [reply proto/regen/data/v1/types.proto:171]: This is not the convention for writing comments. We always start with the name of the field.

File: x/data/client/tx.go (hunk 9 above)
Comment (ryanchristo) [reply x/data/client/tx.go:106]: I don't think the user needs to know the index if they receive the specific iri that is invalid. Also, we've never included an index in an error message as far as I know.

File: x/data/client/tx.go (hunk 8 above)
Comment (ryanchristo) [reply x/data/client/tx.go]: Removed in https://github.com/regen-network/regen-ledger/pull/971/commits/52315dd205bb31753d7d9cf127dd7c493deec89c

File: x/data/server/testsuite/suite.go (hunk 2 above)
Comment (ryanchristo) [reply x/data/server/testsuite/suite.go]: Updated in https://github.com/regen-network/regen-ledger/pull/971/commits/c14f22b14f01fddfd0a8527daf856e9574ed6b8b

File: x/data/client/tx.go (hunk 8 above)
Comment (ryanchristo) [reply x/data/client/tx.go]: Actually this prevents the error from being:

```
invalid iri: failed to parse IRI : regen: prefix required: invalid IRI: invalid request
```

And the error is instead:

```
iri is required
```

**Pull request merged**